lib/WUI/link_content/previews.cpp

```diff
@@ -0,0 +1,63 @@
+#include "previews.h"
+#include "../nhttp/gcode_preview.h"
+
+#include <cstdio>
+
+namespace nhttp::link_content {
+
+using nhttp::printer::GCodePreview;
+using std::nullopt;
+using std::optional;
+using std::string_view;
+using namespace handler;
+
+optional<ConnectionState> Previews::accept(const RequestParser &parser) const {
+    const string_view uri = parser.uri();
+    const constexpr string_view prefix = "/thb/";
+
+    // We claim the /thb/ namespace.
+    // The + 2 is for size specification (s/ or l/)
+    const size_t extra_size = prefix.size() + 2;
+    if (uri.size() < extra_size || uri.substr(0, prefix.size()) != prefix) {
+        return nullopt;
+    }
+
+    // Content of the USB drive is only for authenticated, don't ever try anything without it.
+    if (!parser.authenticated()) {
+        return StatusPage(Status::Unauthorized, parser.can_keep_alive());
+    }
+
+    uint16_t width;
+    uint16_t height;
+
+    if (uri[prefix.size()] == 's') {
+        width = 16;
+        height = 16;
+    } else if (uri[prefix.size()] == 'l') {
+        // TODO: Can we say "something bigger than 16" instead of hardcoding?
```

Comment:
I think we can - ideally return the largest thumbnail availble ... but the gcode filter is not ready to list all of the thumbnails and return the largest one.

Reply:
That means we have to go through the file twice, which is unfortunate :-(.

Anyway, as it needs some kind of extension to the filter, this might want to go to a separate branch some time later?

Reply:
I think we can definitely implement something like that, but since the thumbnails can be stored in arbitrary order, you'd have to read the file twice anyway (detect the list of thumbnails and then pick the largest one). I can think of several optimizations, but let's leave that for later.